File: scripts/export-annotator-json.py
#!/usr/bin/env python3
"""
Export annotator JSON from Postgres (same merged data as web Export button).
No npm/npx needed — uses docker exec into postgres.

Usage on EC2:
  cd /opt/skinfit
  python3 scripts/export-annotator-json.py ~/skinnfit-export.json

Or download to Mac:
  scp -i skinfit-key.pem ubuntu@13.234.166.154:~/skinnfit-export.json ~/Desktop/
"""
from __future__ import annotations

import json
import subprocess
import sys
from datetime import datetime, timezone
from typing import Any

POSTGRES_CONTAINER = "docker-postgres-1"
PG_USER = "skinfit"
PG_DB = "skinfit"

ALL_CATEGORIES = [
    "Active Acne",
    "Acne Scars",
    "Pigmentation",
    "Wrinkles",
    "Sagging & Volume",
    "Under-Eye",
]

GRADE_SCORE = {"A": 1, "B": 2, "C": 3, "D": 4, "E": 5}


def psql_json(query: str) -> Any:
    cmd = [
        "docker",
        "exec",
        POSTGRES_CONTAINER,
        "psql",
        "-U",
        PG_USER,
        "-d",
        PG_DB,
        "-t",
        "-A",
        "-c",
        query,
    ]
    raw = subprocess.check_output(cmd, text=True).strip()
    if not raw or raw == "null":
        return None
    return json.loads(raw)


def normalize_grade(v: Any) -> str:
    if isinstance(v, str) and v in GRADE_SCORE:
        return v
    if isinstance(v, (int, float)) and 1 <= int(v) <= 5:
        return ["A", "B", "C", "D", "E"][int(v) - 1]
    return "A"


def grade_to_score(grade: str) -> int:
    return GRADE_SCORE.get(grade, 1)


def merged_labels_for_export(
    per_user_labels: dict, per_user_shapes: dict, user_sync_at: dict
) -> dict[str, dict[str, dict]]:
    image_indices: set[str] = set()
    for labels in (per_user_labels or {}).values():
        if isinstance(labels, dict):
            image_indices.update(labels.keys())

    out: dict[str, dict[str, dict]] = {}
    for image_key in image_indices:
        best_user = None
        best_ts = 0
        for user_id, labels in (per_user_labels or {}).items():
            if not isinstance(labels, dict) or image_key not in labels:
                continue
            ts = 0
            try:
                from datetime import datetime

                ts = int(
                    datetime.fromisoformat(
                        (user_sync_at or {}).get(user_id, "").replace("Z", "+00:00")
                    ).timestamp()
                    * 1000
                ) if (user_sync_at or {}).get(user_id) else 0
            except Exception:
                ts = 0
            shapes = (per_user_shapes or {}).get(user_id) or []
            has_shapes = any(str(s.get("imageIndex")) == image_key for s in shapes if isinstance(s, dict))
            score = ts + (1 if has_shapes else 0)
            if score >= best_ts:
                best_ts = score
                best_user = user_id
        if best_user:
            out[image_key] = dict((per_user_labels or {}).get(best_user, {}).get(image_key, {}))
    return out


def all_shapes_merged(per_user_shapes: dict) -> list[dict]:
    out: list[dict] = []
    for user_id, shapes in (per_user_shapes or {}).items():
        if not isinstance(shapes, list):
            continue
        for shape in shapes:
            if not isinstance(shape, dict):
                continue
            ann = dict(shape)
            ann["userId"] = user_id
            points = ann.get("points") or []
            min_pts = 2 if ann.get("type") == "line" else 3
            if len(points) >= min_pts:
                out.append(ann)
    return out


def main() -> None:
    out_path = (
        sys.argv[1]
        if len(sys.argv) > 1
        else f"skinnfit-annotations-{datetime.now(timezone.utc).strftime('%Y-%m-%dT%H-%M-%S')}.json"
    )

    state = psql_json(
        "SELECT row_to_json(t) FROM ("
        " SELECT per_user_labels, per_user_shapes, user_sync_at "
        " FROM annotator_state WHERE scope = 'default' LIMIT 1"
        ") t;"
    )
    if not state:
        print("No annotator_state row found", file=sys.stderr)
        sys.exit(1)

    images_rows = psql_json(
        "SELECT coalesce(json_agg(row_to_json(t) ORDER BY sort_order, id), '[]'::json) FROM ("
        " SELECT file_name AS \"fileName\", sort_order AS \"sortOrder\", id "
        " FROM annotator_images ORDER BY sort_order, id"
        ") t;"
    )
    images_rows = images_rows or []

    per_user_labels = state.get("per_user_labels") or {}
    per_user_shapes = state.get("per_user_shapes") or {}
    user_sync_at = state.get("user_sync_at") or {}

    merged_labels = merged_labels_for_export(per_user_labels, per_user_shapes, user_sync_at)
    annotations = []
    for ann in all_shapes_merged(per_user_shapes):
        severity = normalize_grade(ann.get("severity"))
        annotations.append({**ann, "severity": severity, "score": grade_to_score(severity)})

    labels_by_image_index: dict[str, dict] = {}
    for i in range(len(images_rows)):
        merged = {c: {"spec": "", "grade": "A", "score": 1} for c in ALL_CATEGORIES}
        patch = merged_labels.get(str(i)) or {}
        for c in ALL_CATEGORIES:
            entry = patch.get(c) if isinstance(patch, dict) else None
            if isinstance(entry, dict):
                grade = normalize_grade(entry.get("grade") or entry.get("score"))
                merged[c] = {
                    "spec": entry.get("spec") or "",
                    "grade": grade,
                    "score": grade_to_score(grade),
                }
        labels_by_image_index[str(i)] = merged

    payload = {
        "schemaVersion": 2,
        "app": "skinnfit-clinical-annotator",
        "exportedAt": datetime.now(timezone.utc).isoformat(),
        "note": (
            "Images are not embedded. Match `images[].fileName` to files on disk. "
            "Points are normalized 0–1 vs image width/height. `grade` is A–E (A=least severe); "
            "`score` is numeric 1–5 for eval pipelines. Merged export includes all collaborators' shapes and labels."
        ),
        "imageCount": len(images_rows),
        "images": [
            {
                "index": i,
                "fileName": row.get("fileName") or f"image-{i + 1}",
                "imageWidth": None,
                "imageHeight": None,
            }
            for i, row in enumerate(images_rows)
        ],
        "labelsByImageIndex": labels_by_image_index,
        "annotations": annotations,
    }

    text = json.dumps(payload, indent=2)
    with open(out_path, "w", encoding="utf-8") as f:
        f.write(text)
    mb = len(text) / 1024 / 1024
    print(
        f"Exported {len(annotations)} annotations, {len(images_rows)} images → {out_path} ({mb:.2f} MB)",
        file=sys.stderr,
    )


if __name__ == "__main__":
    main()

File: src/scripts/export-annotator-json.ts
/**
 * Export annotator JSON from the database (same merged data as the web Export button).
 * No browser needed — reads Postgres directly.
 *
 * On EC2 (from laptop):
 *   ssh -i skinfit-key.pem ubuntu@13.234.166.154 \
 *     'cd /opt/skinfit && LOCAL_POSTGRES_URL="postgresql://skinfit:skinfit_local_dev@127.0.0.1:5433/skinfit" \
 *      npx tsx src/scripts/export-annotator-json.ts /tmp/skinnfit-export.json && cat /tmp/skinnfit-export.json' \
 *     > skinnfit-export.json
 *
 * On EC2 (already SSH'd):
 *   cd /opt/skinfit
 *   LOCAL_POSTGRES_URL='postgresql://skinfit:skinfit_local_dev@127.0.0.1:5433/skinfit' \
 *     npx tsx src/scripts/export-annotator-json.ts ~/skinnfit-export.json
 */
import "dotenv/config";
import { writeFileSync } from "fs";
import { asc, eq } from "drizzle-orm";
import { db } from "@/src/db";
import { annotatorImages, annotatorState } from "@/src/db/schema";
import {
  ANNOTATOR_SCOPE,
  allShapesMerged,
  mergedLabelsForExport,
  parseCollaborationStore,
} from "@/src/lib/annotatorCollaboration";
import {
  normalizeSeverityGrade,
  severityGradeToScore,
  type SeverityGrade,
} from "@/src/lib/annotatorSeverityGrade";

const ALL_CATEGORIES = [
  "Active Acne",
  "Acne Scars",
  "Pigmentation",
  "Wrinkles",
  "Sagging & Volume",
  "Under-Eye",
] as const;

type Category = (typeof ALL_CATEGORIES)[number];

function defaultEntry(cat: Category): { spec: string; grade: SeverityGrade } {
  return { spec: "", grade: "A" };
}

function normalizeCategoryEntry(raw: {
  spec?: string;
  grade?: unknown;
  score?: unknown;
}): { spec: string; grade: SeverityGrade } {
  return {
    spec: typeof raw.spec === "string" ? raw.spec : "",
    grade: normalizeSeverityGrade(raw.grade ?? raw.score, "A"),
  };
}

async function main() {
  const outPath =
    process.argv[2] ??
    `skinnfit-annotations-${new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19)}.json`;

  const [stateRow] = await db
    .select()
    .from(annotatorState)
    .where(eq(annotatorState.scope, ANNOTATOR_SCOPE))
    .limit(1);

  const store = parseCollaborationStore(stateRow ?? null);
  const mergedLabels = mergedLabelsForExport(store);
  const rawShapes = allShapesMerged(store);

  const annotations = rawShapes
    .filter((ann) => Array.isArray(ann.points) && ann.points.length >= (ann.type === "line" ? 2 : 3))
    .map((ann) => {
      const severity = normalizeSeverityGrade(ann.severity, "A");
      return {
        ...ann,
        severity,
        score: severityGradeToScore(severity),
      };
    });

  const images = await db
    .select({
      fileName: annotatorImages.fileName,
      sortOrder: annotatorImages.sortOrder,
    })
    .from(annotatorImages)
    .orderBy(asc(annotatorImages.sortOrder), asc(annotatorImages.id));

  const labelsByImageIndex: Record<
    string,
    Record<Category, { spec: string; grade: SeverityGrade; score: number }>
  > = {};

  for (let i = 0; i < images.length; i++) {
    const base = Object.fromEntries(ALL_CATEGORIES.map((c) => [c, defaultEntry(c)])) as Record<
      Category,
      { spec: string; grade: SeverityGrade }
    >;
    const patch = mergedLabels[String(i)] ?? {};
    const merged = { ...base };
    for (const c of ALL_CATEGORIES) {
      const entry = patch[c];
      if (entry) merged[c] = normalizeCategoryEntry(entry);
    }
    labelsByImageIndex[String(i)] = Object.fromEntries(
      ALL_CATEGORIES.map((c) => [
        c,
        {
          spec: merged[c].spec,
          grade: merged[c].grade,
          score: severityGradeToScore(merged[c].grade),
        },
      ])
    ) as Record<Category, { spec: string; grade: SeverityGrade; score: number }>;
  }

  const payload = {
    schemaVersion: 2,
    app: "skinnfit-clinical-annotator",
    exportedAt: new Date().toISOString(),
    note:
      "Images are not embedded. Match `images[].fileName` to files on disk. Points are normalized 0–1 vs image width/height. `grade` is A–E (A=least severe); `score` is numeric 1–5 for eval pipelines. Merged export includes all collaborators' shapes and labels.",
    imageCount: images.length,
    images: images.map((img, i) => ({
      index: i,
      fileName: img.fileName,
      imageWidth: null,
      imageHeight: null,
    })),
    labelsByImageIndex,
    annotations,
  };

  const json = JSON.stringify(payload, null, 2);
  writeFileSync(outPath, json, "utf8");
  console.error(
    `Exported ${annotations.length} annotations, ${images.length} images → ${outPath} (${(json.length / 1024 / 1024).toFixed(2)} MB)`
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
